File: app/manager.js
const axios = require('axios');
const LoadingSpinner = require('./ui/spinner');
const ProgressBar = require('./ui/progress');
const { gradient } = require('./ui/gradient');

const DEFAULT_API_BASE_URL = 'https://minhdong.site/api/v1/facebook';

class FacebookPageManager {
  constructor(apiKey) {
    const baseURL = process.env.API_BASE_URL || DEFAULT_API_BASE_URL;
    this.client = axios.create({
      baseURL,
      headers: {
        'X-API-Key': apiKey,
        'Content-Type': 'application/json'
      },
      timeout: 60000
    });
    this.stats = {
      total: 0,
      success: 0,
      failed: 0,
      skipped: 0
    };
  }

  async regAndTransfer(quantity, config) {
    const { cookieGoc, passGoc, uidNhan, cookieNhan } = config;

    console.log('\n' + gradient('+----------------------------------------------------------+'));
    console.log(gradient('|   REG & TRANSFER PAGE AUTOMATION                         |'));
    console.log(gradient('+----------------------------------------------------------+') + '\n');
    console.log(`So luong Page can tao: ${quantity}`);
    console.log(`UID nhan: ${uidNhan}\n`);

    this.stats = { total: quantity, success: 0, failed: 0, skipped: 0 };
    const progressBar = new ProgressBar(quantity, 'Tien trinh');

    for (let i = 1; i <= quantity; i++) {
      console.log(`\n${'='.repeat(60)}`);
      console.log(gradient(`[${i}/${quantity}] Bat dau quy trinh...`));
      console.log('='.repeat(60));

      try {
        const spinner1 = new LoadingSpinner('Đang tạo Page mới');
        spinner1.start();

        await this._delay(1000);
        const regRes = await this.client.post('/page/create-cookie', {
          cookie: cookieGoc,
          use_random_name: true,
          use_random_avatar: true,
          use_random_bio: true
        });

        if (!regRes.data.success) {
          spinner1.fail(`Tao Page that bai: ${regRes.data.error || 'Loi khong xac dinh'}`);
          this.stats.failed++;
          progressBar.update(i, '[FAIL]');
          continue;
        }

        const newPage = {
          profile_plus_id: regRes.data.profile_plus_id,
          name: regRes.data.name,
          page_cookie: `${cookieGoc}; i_user=${regRes.data.profile_plus_id}`
        };

        spinner1.succeed(`Tao Page thanh cong: ${newPage.name}`);
        console.log(`   Page ID: ${newPage.profile_plus_id}`);

        const transferSuccess = await this._processSinglePage(newPage, passGoc, uidNhan, cookieNhan);

        if (transferSuccess) {
          this.stats.success++;
          progressBar.update(i, '[OK]');
        } else {
          this.stats.failed++;
          progressBar.update(i, '[FAIL]');
        }

        if (i < quantity) {
          const delayTime = 120000;
          const spinner2 = new LoadingSpinner(`Cho ${delayTime / 1000}s truoc khi tiep tuc`);
          spinner2.start();
          await this._delay(delayTime);
          spinner2.info('Tiep tuc...');
        }

      } catch (err) {
        spinner1.stop();
        console.error(`Loi nghiem trong tai buoc ${i}:`, err.message);
        this.stats.failed++;
        progressBar.update(i, '[FAIL]');
      }
    }

    this._displayFinalReport();
  }

  async _processSinglePage(page, passGoc, uidNhan, cookieNhan) {
    const { profile_plus_id, name, page_cookie } = page;

    try {
      console.log('\n  [1/3] Gui loi moi Admin...');
      const spinner1 = new LoadingSpinner('  Đang xử lý lời mời');
      spinner1.start();

      const addRes = await this.client.post('/page/add-permission-cookie', {
        cookie: page_cookie,
        target_user_id: uidNhan,
        password: passGoc
      });

      if (!addRes.data.success) {
        spinner1.fail(`Loi gui loi moi: ${addRes.data.error}`);
        return false;
      }

      const inviteId = addRes.data.invite_id;
      spinner1.succeed(`Gui loi moi thanh cong (ID: ${inviteId})`);

      console.log('  [2/3] Chap nhan loi moi...');
      const spinner2 = new LoadingSpinner('  Nick nhận đang xử lý');
      spinner2.start();

      const acceptRes = await this.client.post('/page/accept-invitation-cookie', {
        cookie: cookieNhan,
        profile_admin_invite_id: inviteId,
        user_id: profile_plus_id,
        accept: true
      });

      if (!acceptRes.data.success) {
        spinner2.fail(`Loi chap nhan: ${acceptRes.data.error}`);
        return false;
      }

      spinner2.succeed('Chap nhan loi moi thanh cong');

      const uidGoc = this._extractUid(page_cookie);
      console.log('  [3/3] Go quyen nick goc...');
      const spinner3 = new LoadingSpinner(`  Đang gỡ admin ${uidGoc}`);
      spinner3.start();

      const removeRes = await this.client.post('/page/remove-admin-cookie', {
        cookie: page_cookie,
        admin_id: uidGoc,
        password: passGoc,
        profile_plus_id: profile_plus_id
      });

      if (removeRes.data.success) {
        spinner3.succeed('Go quyen thanh cong');
        console.log(`\n  Hoan tat ban giao: ${name}`);
        return true;
      } else {
        spinner3.fail(`Loi go quyen: ${removeRes.data.error}`);
        return false;
      }

    } catch (error) {
      console.error(`  Loi trong qua trinh transfer:`, error.response?.data || error.message);
      return false;
    }
  }

  _displayFinalReport() {
    console.log('\n\n' + gradient('+----------------------------------------------------------+'));
    console.log(gradient('|                    BAO CAO KET QUA                       |'));
    console.log(gradient('+----------------------------------------------------------+'));

    const successRate = ((this.stats.success / this.stats.total) * 100).toFixed(1);

    console.log(`\nThong ke:`);
    console.log(`   Tong so: ${this.stats.total}`);
    console.log(`   Thanh cong: ${this.stats.success}`);
    console.log(`   That bai: ${this.stats.failed}`);
    console.log(`   Ty le thanh cong: ${successRate}%`);

    console.log(`\nThoi gian hoan thanh: ${new Date().toLocaleString('vi-VN')}\n`);
  }

  _extractUid(cookie) {
    const match = cookie.match(/c_user=(\d+)/);
    return match ? match[1] : null;
  }

  _delay(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }

  _getRandomDelay(min, max) {
    return Math.floor(Math.random() * (max - min + 1)) + min;
  }
}

module.exports = FacebookPageManager;

File: python/manager.py
from __future__ import annotations

import os
import time
from dataclasses import dataclass, field
from typing import Dict

import requests

from .ui.gradient import gradient
from .ui.progress import ProgressBar
from .ui.spinner import LoadingSpinner


DEFAULT_API_BASE_URL = "https://minhdong.site/api/v1/facebook"


@dataclass
class Stats:
    total: int = 0
    success: int = 0
    failed: int = 0
    skipped: int = 0


class FacebookPageManager:
    def __init__(self, api_key: str) -> None:
        base_url = os.getenv("API_BASE_URL") or DEFAULT_API_BASE_URL
        self.client = requests.Session()
        self.client.headers.update(
            {
                "X-API-Key": api_key,
                "Content-Type": "application/json",
            }
        )
        self.base_url = base_url.rstrip("/")
        self.stats: Stats = Stats()

    def _post(self, path: str, json: Dict) -> Dict:
        resp = self.client.post(f"{self.base_url}{path}", json=json, timeout=60)
        resp.raise_for_status()
        return resp.json()

    def reg_and_transfer(self, quantity: int, config: Dict[str, str]) -> None:
        cookie_goc = config["cookieGoc"]
        pass_goc = config["passGoc"]
        uid_nhan = config["uidNhan"]
        cookie_nhan = config["cookieNhan"]

        print("\n" + gradient("+----------------------------------------------------------+"))
        print(gradient("|   REG & TRANSFER PAGE AUTOMATION                         |"))
        print(gradient("+----------------------------------------------------------+") + "\n")
        print(f"So luong Page can tao: {quantity}")
        print(f"UID nhan: {uid_nhan}\n")

        self.stats = Stats(total=quantity, success=0, failed=0, skipped=0)
        progress_bar = ProgressBar(quantity, "Tien trinh")

        for i in range(1, quantity + 1):
            print(f"\n{'=' * 60}")
            print(gradient(f"[{i}/{quantity}] Bat dau quy trinh..."))
            print("=" * 60)

            spinner1 = LoadingSpinner("Đang tạo Page mới")
            try:
                spinner1.start()
                time.sleep(1.0)
                reg_res = self._post(
                    "/page/create-cookie",
                    {
                        "cookie": cookie_goc,
                        "use_random_name": True,
                        "use_random_avatar": True,
                        "use_random_bio": True,
                    },
                )

                if not reg_res.get("success"):
                    spinner1.fail(
                        f"Tao Page that bai: {reg_res.get('error') or 'Loi khong xac dinh'}"
                    )
                    self.stats.failed += 1
                    progress_bar.update(i, "[FAIL]")
                    continue

                new_page = {
                    "profile_plus_id": reg_res.get("profile_plus_id"),
                    "name": reg_res.get("name"),
                    "page_cookie": f"{cookie_goc}; i_user={reg_res.get('profile_plus_id')}",
                }

                spinner1.succeed(f"Tao Page thanh cong: {new_page['name']}")
                print(f"   Page ID: {new_page['profile_plus_id']}")

                transfer_success = self._process_single_page(
                    new_page, pass_goc, uid_nhan, cookie_nhan
                )

                if transfer_success:
                    self.stats.success += 1
                    progress_bar.update(i, "[OK]")
                else:
                    self.stats.failed += 1
                    progress_bar.update(i, "[FAIL]")

                if i < quantity:
                    delay_time = 120_000
                    spinner2 = LoadingSpinner(
                        f"Cho {delay_time // 1000}s truoc khi tiep tuc"
                    )
                    spinner2.start()
                    self._delay(delay_time / 1000.0)
                    spinner2.info("Tiep tuc...")

            except Exception as exc:
                spinner1.stop()
                print(f"Loi nghiem trong tai buoc {i}:", exc)
                self.stats.failed += 1
                progress_bar.update(i, "[FAIL]")

        self._display_final_report()

    def _process_single_page(
        self,
        page: Dict[str, str],
        pass_goc: str,
        uid_nhan: str,
        cookie_nhan: str,
    ) -> bool:
        profile_plus_id = page["profile_plus_id"]
        name = page["name"]
        page_cookie = page["page_cookie"]

        try:
            print("\n  [1/3] Gui loi moi Admin...")
            spinner1 = LoadingSpinner("  Đang xử lý lời mời")
            spinner1.start()
            add_res = self._post(
                "/page/add-permission-cookie",
                {
                    "cookie": page_cookie,
                    "target_user_id": uid_nhan,
                    "password": pass_goc,
                },
            )
            if not add_res.get("success"):
                spinner1.fail(f"Loi gui loi moi: {add_res.get('error')}")
                return False
            invite_id = add_res.get("invite_id")
            spinner1.succeed(f"Gui loi moi thanh cong (ID: {invite_id})")

            print("  [2/3] Chap nhan loi moi...")
            spinner2 = LoadingSpinner("  Nick nhận đang xử lý")
            spinner2.start()
            accept_res = self._post(
                "/page/accept-invitation-cookie",
                {
                    "cookie": cookie_nhan,
                    "profile_admin_invite_id": invite_id,
                    "user_id": profile_plus_id,
                    "accept": True,
                },
            )
            if not accept_res.get("success"):
                spinner2.fail(f"Loi chap nhan: {accept_res.get('error')}")
                return False
            spinner2.succeed("Chap nhan loi moi thanh cong")

            uid_goc = self._extract_uid(page_cookie)
            print("  [3/3] Go quyen nick goc...")
            spinner3 = LoadingSpinner(f"  Đang gỡ admin {uid_goc}")
            spinner3.start()
            remove_res = self._post(
                "/page/remove-admin-cookie",
                {
                    "cookie": page_cookie,
                    "admin_id": uid_goc,
                    "password": pass_goc,
                    "profile_plus_id": profile_plus_id,
                },
            )
            if remove_res.get("success"):
                spinner3.succeed("Go quyen thanh cong")
                print(f"\n  Hoan tat ban giao: {name}")
                return True
            spinner3.fail(f"Loi go quyen: {remove_res.get('error')}")
            return False

        except Exception as exc:
            print("  Loi trong qua trinh transfer:", exc)
            return False

    def _display_final_report(self) -> None:
        print("\n\n" + gradient("+----------------------------------------------------------+"))
        print(gradient("|                    BAO CAO KET QUA                       |"))
        print(gradient("+----------------------------------------------------------+"))
        success_rate = (
            (self.stats.success / self.stats.total) * 100 if self.stats.total else 0.0
        )
        print("\nThong ke:")
        print(f"   Tong so: {self.stats.total}")
        print(f"   Thanh cong: {self.stats.success}")
        print(f"   That bai: {self.stats.failed}")
        print(f"   Ty le thanh cong: {success_rate:.1f}%")
        print(f"\nThoi gian hoan thanh: {time.strftime('%Y-%m-%d %H:%M:%S')}\n")

    @staticmethod
    def _extract_uid(cookie: str) -> str | None:
        marker = "c_user="
        idx = cookie.find(marker)
        if idx == -1:
            return None
        start = idx + len(marker)
        end = cookie.find(";", start)
        return cookie[start:end] if end != -1 else cookie[start:]

    @staticmethod
    def _delay(seconds: float) -> None:
        time.sleep(seconds)

